certora/spec/CVL2/timelockAuthorizerTed.spec

```diff
@@ -0,0 +1,556 @@
+        "Pending root changed root to somebody else.";
+    // if the function f changed the root, then f must be claimRoot
+    assert rootBefore != rootAfter =>
+        f.selector == sig:claimRoot().selector,
+        "Root changed by a function other than claimRoot.";
+}
+
+
+// STATUS - verified
+// Only two functions can change pending root: claimRoot and setPendingRoot.
+// claimRoot can only be called by the execution helper.
+// setPendingRoot can only be called by the root or the execution helper.
+// https://vaas-stg.certora.com/output/40577/7efc90865f0f47d09d3d42dab4214400/?anonymousKey=26431befd59fbb32ae6201c42e7f456e7ccceb06
+rule pendingRootChangesOnlyWithSetPendingRootOrClaimRoot(env e, method f) {
+    address pendingRootBefore = getPendingRoot();
+    address executionHelper = getTimelockExecutionHelper();
+
+    // Invoke any function
+    calldataarg args;
+    f(e, args);
+
+    address pendingRootAfter = getPendingRoot();
+
+    // if the function f changed the root, then f must be setPendingRoot or claimRoot.
+    assert pendingRootBefore != pendingRootAfter =>
+        ( f.selector == sig:setPendingRoot(address).selector ||
+        f.selector == sig:claimRoot().selector ),
+        "Pending root changed by a function other than setPendingRoot or claimRoot.";
+    // if the function changed the pending root, the sender was pending root or execution helper.
+    assert pendingRootBefore != pendingRootAfter =>
+        e.msg.sender == pendingRootBefore ||
+        e.msg.sender == executionHelper,
+        "Pending root changed by somebody, who was not root.";
+    assert (pendingRootBefore == 0 && pendingRootAfter != 0) => e.msg.sender == executionHelper,
+        "pendingRoot can only be changed by the executer";
+}
+
+
+// STATUS - verified
+// ScheduledExecution that has already been cancelled or executed
+// cannot be canceled again.
+// https://prover.certora.com/output/40577/0a8cadb54810435c859098750efa8ee0/?anonymousKey=58e33df6c4063dd4882ec3f40d8b86d816951102
+rule scheduledExecutionCanBeCancelledOnlyOnce(env e, uint256 index) {
+    bool canceled_before = getSchedExeCancelled(index);
+    bool executed_before = getSchedExeExecuted(index);
+
+    cancel@withrevert(e, index);
+    assert (canceled_before || executed_before) => lastReverted;
+}
+
+
+// STATUS - verified
+// ScheduledExecution that has already been cancelled or executed
+// cannot be executed again.
+// https://prover.certora.com/output/40577/0a8cadb54810435c859098750efa8ee0/?anonymousKey=58e33df6c4063dd4882ec3f40d8b86d816951102
+rule scheduledExecutionCanBeExecutedOnlyOnce(env e, uint256 index) {
+    bool canceled_before = getSchedExeCancelled(index);
+    bool executed_before = getSchedExeExecuted(index);
+
+    execute@withrevert(e, index);
+    assert (canceled_before || executed_before) => lastReverted;
+}
+
+
+// STATUS - verified
+// Scheduled execution can only be executed by those supplied as executors during scheduling
```

Comment:
The description seems to be misleading. Rule name better describes the goal of the rule.

Reply:
I thought that we do not check for things, that are in require in customers code. And as `require(isExecutor(sender))` is in the `execute` function, the description is equivalent with the following one, that I replated it with:
"User cannot become an executor for already scheduled execution, meaning that
a user has to be supplied as one of the executor when scheduling an execution,
to be able to execute this execution."
I try to make the descriptions as something, that adds some informative value to the name of the rule. Having descriptions, that only repeat the information already present in the name of the rule is a bit redundant, don't you think?

Reply:
I thought that we do not check for things, that are in require in customer's code. And as `require(isExecutor(sender))` is in the `execute` function, the description is equivalent to the following one, which I replated it with:
"User cannot become an executor for already scheduled execution, meaning that
a user has to be supplied as one of the executors when scheduling an execution,
to be able to execute this execution."
I try to make the descriptions as something, that adds some informative value to the name of the rule. Having descriptions, that only repeat the information already present in the name of the rule is a bit redundant, right?

Reply:
1. if, for example, revert condition is stated in doc/function description, then it's not that redundant to check it because this is a logic developers want to see.
2. according to your current description, I expect that the rule will be implemented like:
```
bool isExecutor = isExecutor(user);
execute@withrevert();
assert lastReverted => !isExecutor;
```
Implementation on the other hand reflects the name

Reply:
By the current description you mean the one in my comment above?

Reply:
I mean the one in the spec